ghost/email-service/lib/mailgun-email-provider.js

```diff
@@ -0,0 +1,168 @@
+                def.token,
+                `%recipient.${def.id}%`
+            );
+        }
+        return data;
+    }
+
+    /**
+     * Create mailgun error message for storing in the database
+     * @param {Object} error
+     * @param {string} error.message
+     * @param {string} error.details
+     * @returns {string}
+    */
+    #createMailgunErrorMessage(error) {
+        const message = (error?.message || '') + ':' + (error?.details || '');
+        return message.slice(0, 2000);
+    }
+
+    /**
+     * Send an email using the Mailgun API
+     * @param {import('./sending-service').EmailData} data
+     * @param {EmailSendingOptions} options
+     * @returns {Promise<EmailProviderSuccessResponse>}
+    */
+    async send(data, options) {
+        const {
+            subject,
+            html,
+            plaintext,
+            from,
+            replyTo,
+            emailId,
+            recipients,
+            replacementDefinitions
+        } = data;
+
+        logging.info(`Sending email to ${recipients.length} recipients`);
+        const startTime = Date.now();
+        debug(`sending message to ${recipients.length} recipients`);
+
+        try {
+            const messageData = {
+                subject,
+                html,
+                plaintext,
+                from,
+                replyTo,
+                id: emailId,
+                track_opens: !!options.openTrackingEnabled,
+                track_clicks: !!options.clickTrackingEnabled
+            };
+
+            // create recipient data for Mailgun using replacement definitions
+            const recipientData = recipients.reduce((acc, recipient) => {
+                acc[recipient.email] = this.#createRecipientData(recipient.replacements);
+                return acc;
+            }, {});
+
+            // update content to use Mailgun variable syntax for all replacements
+            ['html', 'plaintext', 'subject'].forEach((key) => {
+                if (messageData[key]) {
+                    messageData[key] = this.#updateRecipientVariables(messageData[key], replacementDefinitions);
+                }
+            });
+
+            // send the email using Mailgun
+            // uses empty replacements array as we've already replaced all tokens with Mailgun variables
+            const response = await this.#mailgunClient.send(
+                messageData,
+                recipientData,
+                []
+            );
+
+            debug(`sent message (${Date.now() - startTime}ms)`);
+            logging.info(`Sent message (${Date.now() - startTime}ms)`);
+
+            // Return mailgun provider id, trim <> from response
+            return {
+                id: response.id.trim().replace(/^<|>$/g, '')
+            };
+        } catch ({error, messageData}) {
+            // REF: possible mailgun errors https://documentation.mailgun.com/en/latest/api-intro.html#status-codes
+            let ghostError = new errors.EmailError({
```

Comment:
@SimonBackx Have repurposed existing `EmailError` here for passing the error instead of creating new `EmailProviderError` as it seemed unnecessary, but i am not 100% on that, wdyt ?

Reply:
Should be fine 👌